src/textual/widgets/_data_table.py

```diff
@@ -2107,13 +2107,15 @@ def _get_fixed_offset(self) -> Spacing:
 
     def sort(
         self,
-        *columns: ColumnKey | str,
+        by: Iterable[ColumnKey | str] | Callable,
```

Comment:
Rather than a breaking change, could we keep the positional arguments, but add a `key` argument. This feels like it would be closer to sorting lists, and would be less astonishing to the dev.

If both columns and a key function are specified, then I would think that would be an error.

Reply:
> If both columns and a key function are specified, then I would think that would be an error.

This feels confusing to me to be honest - it means we'll have a method with two parameters controlling the same thing (how the data is sorted) that can't be supplied together. If we keep them as a single param then the method signature/typing conveys that they can't both be supplied. By having separate params, we're opening up an avenue for people to use the API incorrectly which didn't exist before. If we really want to avoid a breaking change my vote would be for a new method.

Reply:
Let me explore how this change may work.

With this change:

`table.sort()`  Sort by default column order.
`table.sort(("foo",))`   Sort by single column (feels clumsy, even if it was a list and not a tuple of one.
`table.sort(("foo", "bar"))`  Still feels clumsy to me. The double nesting of parenthesis would be a double take.
`table.sort(my_key)`  # Sort by key, clear enough.

With positional args, and key parameter:

`table.sort()`
`table.sort("foo")`  I think this is clear that it sorts one column.
`table.sort("foo", "bar")`  Sorts two columns
`table.sort(key=my_key)`  Sorts by key.

I hear what you are saying about the ambiguity of specifying both columns and a key.

How about, if you specify columns and a key, then the key gets a tuple of the specified columns, and not the entire row.

I feel this is actually useful, and not a gimmick to get around @darrenburns objection.

It will allow us to have a key function that isn't tied to the number of columns. At the moment, if we add a column or change their order, our key function could become outdated. Consider this:

```python
def sort_foo_bar(row:tuple[str, str]) -> str:
    return "".join(row).lower()

table.sort("foo", "bar", key=sort_foo_bar)
```

Since we have specified the columns for the key function, it won't change the behaviour if we change the order of the columns, or add more columns.

Thoughts?

Reply:
I think that's a really good idea, I like it - good point about the key function potentially becoming outdated after adding columns too 👍

Reply:
@willmcgugan, I like this solution as well and can get it implemented. Thanks!